oshdb-util/src/main/java/org/heigit/bigspatialdata/oshdb/util/geometry/OSHDBGeometryBuilder.java

```diff
@@ -285,14 +285,95 @@ private static Geometry getMultiPolygonGeometry(
     }
-    if (touchingRings) {
-      // try to clean up geometry by calling buffer(0).
-      // see https://locationtech.github.io/jts/jts-faq.html#G1
-      result = result.buffer(0);
-    }
     return result;
   }
 
+  private static Stream<Stream<OSMNode>> waysToLines(
+      Stream<OSMEntity> members, OSHDBTimestamp timestamp) {
+    return members
+        .map(osm -> (OSMWay) osm)
+        .filter(Objects::nonNull)
+        .filter(OSMEntity::isVisible)
+        .map(way -> way.getRefEntities(timestamp)
+            .filter(Objects::nonNull)
+            .filter(OSMEntity::isVisible)
+        );
+  }
+
+  private static void mergeTouchingRings(List<LinkedList<OSMNode>> ringsNodes) {
+    Map<Segment, LinkedList<OSMNode>> ringSegments = new HashMap<>();
+    List<LinkedList<OSMNode>> mergedRings = new LinkedList<>();
+    for (LinkedList<OSMNode> ringNodes : ringsNodes) {
+      List<Segment> thisRingSegments = new ArrayList<>(ringNodes.size() - 1);
+      int numNodes = ringNodes.size();
+      long prevNodeId = ringNodes.get(0).getId();
+      for (int i = 1; i < numNodes; i++) {
+        long thisNodeId = ringNodes.get(i).getId();
+        Segment segment = new Segment(prevNodeId, thisNodeId);
+        if (!ringSegments.containsKey(segment)) {
+          thisRingSegments.add(segment);
+        } else {
+          // merge this ring with the previous one
+          LinkedList<OSMNode> targetNodes = ringSegments.get(segment);
+          ringSegments.values().remove(targetNodes);
+          cutAtSegment(targetNodes, segment);
+          cutAtSegment(ringNodes, segment);
+          mergeSegmentsToRing(targetNodes, ringNodes);
+          // clean up
+          // add merged segments to thisRingSements
+          thisRingSegments.clear();
+          for (int j = 1; j < targetNodes.size(); j++) {
+            thisRingSegments.add(
+                new Segment(targetNodes.get(j - 1).getId(), targetNodes.get(j).getId())
+            );
+          }
+          // mark merged ring as to be removed
+          mergedRings.add(ringNodes);
+          ringNodes = targetNodes;
+          break;
+        }
+        prevNodeId = thisNodeId;
+      }
+      for (Segment thisRingSegment : thisRingSegments) {
+        ringSegments.put(thisRingSegment, ringNodes);
+      }
+    }
+    ringsNodes.removeAll(mergedRings);
+  }
+
+  private static void cutAtSegment(LinkedList<OSMNode> ring, Segment segment) {
+    ring.remove(0);
+    while (true) {
+      if (ring.getFirst().getId() == segment.id1 && ring.getLast().getId() == segment.id2
+          || ring.getFirst().getId() == segment.id2 && ring.getLast().getId() == segment.id1) {
+        return;
+      }
+      ring.add(ring.removeFirst());
+    }
+  }
+
+  private static void mergeSegmentsToRing(LinkedList<OSMNode> target, LinkedList<OSMNode> source) {
+    if (target.getFirst().getId() == source.getFirst().getId()) {
+      Collections.reverse(source);
```

Comment:
you mean having four `while` loops (two here and the current two in the `else` branch) to check all possible joining possibilities? that would be more performant, I guess :+1:

Reply:
ok, I took a second look. "some" reversing the list is necessary anyway, because in order for the two lines `target`/`source` to be joined into a proper ring at the end of this method, they need to have the correct order anyway. One might save a few swap operations by doing this at the end of the method, but I believe it's at this point an unnecessary optimization and not worth the added complexity in the code.